src/main/java/io/cdap/plugin/gcp/gcs/sink/GCSBatchSink.java

```diff
@@ -82,18 +111,95 @@ public void prepareRun(BatchSinkContext context) throws Exception {
+
   }
 
   @Override
   protected Map<String, String> getFileSystemProperties(BatchSinkContext context) {
     return GCPUtils.getFileSystemProperties(config, config.getPath(), new HashMap<>());
   }
 
+  @Override
+  protected String getOutputDir(long logicalStartTime) {
+    this.outputPath = super.getOutputDir(logicalStartTime);
+    return this.outputPath;
+  }
+
   @Override
   protected void recordLineage(LineageRecorder lineageRecorder, List<String> outputFields) {
     lineageRecorder.recordWrite("Write", "Wrote to Google Cloud Storage.", outputFields);
   }
 
+  @Override
+  public void onRunFinish(boolean succeeded, BatchSinkContext context) {
+    super.onRunFinish(succeeded, context);
+    recordMetric(succeeded, context);
+  }
+
+  private void recordMetric(boolean succeeded, BatchSinkContext context) {
+    if (!succeeded) {
+      return;
+    }
+
+    try {
+      recordMetricForAllFiles(context);
+    } catch (Exception e) {
+      LOG.warn("Metrics for the number of affected rows maybe incorrect.", e);
+    }
+  }
+
+  private void recordMetricForAllFiles(BatchSinkContext context) throws IOException {
+    GCSPath gcsPath = GCSPath.from(outputPath);
+    //add all the configurations necessary to create file system
+    Configuration configuration = new Configuration();
+    getFileSystemProperties(context).entrySet().stream()
+      .forEach(entry -> configuration.set(entry.getKey(), entry.getValue()));
+
+    try (FileSystem fileSystem = getFileSystem(outputPath, configuration)) {
+      if (!(fileSystem instanceof GoogleHadoopFileSystem)) {
+        return;
+      }
+
+      GoogleCloudStorageFileSystem gcsFs = ((GoogleHadoopFileSystem) fileSystem).getGcsFs();
+      try {
+        //For each file, read metadata to get count and sum them up
+        List<FileInfo> fileInfos = gcsFs.listAllFileInfoForPrefix(gcsPath.getUri());
+        fileInfos.stream()
+          .forEach(fileInfo -> recordMetric(context.getMetrics(), extractRecordCount(fileInfo.getAttributes())));
+      } finally {
+        gcsFs.close();
+      }
+
+    }
+  }
+
+  private void recordMetric(StageMetrics metrics, long totalRows) {
+    if (totalRows == 0) {
+      return;
+    }
+    //work around since StageMetrics count() only takes int as of now
+    int cap = 10000; // so the loop will not cause significant delays
+    long count = totalRows / Integer.MAX_VALUE;
+    if (count > cap) {
+      LOG.warn("Total record count is too high! Metric for the number of affected rows may not be updated correctly");
+    }
+    count = count < cap ? count : cap;
+    for (int i = 0; i <= count && totalRows > 0; i++) {
+      int rowCount = totalRows < Integer.MAX_VALUE ? (int) totalRows : Integer.MAX_VALUE;
+      metrics.count(RECORDS_UPDATED_METRIC, rowCount);
+      totalRows -= rowCount;
+    }
+  }
+
+  private long extractRecordCount(Map<String, byte[]> metadata) {
+    byte[] value = metadata.getOrDefault(RECORD_COUNT, Long.valueOf(0).toString().getBytes());
```

Comment:
```String.valueOf(0L).getBytes(StandardCharsets.UTF_8);```

Always provide a Charset when calling getBytes on String.

Reply:
Also, seems simpler if you do:

```
byte[] value = metadata.get(RECORD_COUNT);
return value == null ? 0L : Long.parseLong(new String(value, StandardCharsets.UTF_8));
```

Reply:
Done